### ISSUE TITLE
nesting media conditions — example from the CSS spec fails?

### ISSUE THREAD

**Issue by revelt** [csstree/csstree#118]
CSS spec says:

> Media conditions can be grouped by wrapping them in parentheses () which can then be nested within a condition the same as a single media query. For example, `(not (color)) or (hover)` is true on devices that are monochrome and/or that have hover capabilities.
https://www.w3.org/TR/mediaqueries-4/#media-conditions

When I try to assemble a test media query with `(not (color)) or (hover)`, the [validator](https://csstree.github.io/docs/validator.html) complains:

```
@media (not (color)) or (hover) {}
```

<img width="437" alt="parser2" src="https://user-images.githubusercontent.com/8344688/72510627-af1a7d80-3841-11ea-9a31-36cdf3711fa9.png">

Maybe I misunderstood the spec but if you can, please clarify, is it a parser's problem or is it something I misunderstoon in the spec. I'm looking to truly understand the part `(not (color)) or (hover)`.

**Comment by lahmatiy** [comment]
Your example is correct and parser should not fail in this case. I can reproduce the problem. Definitely that's a bug, I will investigate the issue.

**Comment by revelt** [comment]
It's interesting, Chrome doesn't like queries conditions in brackets. I changed `color` to `monochrome` but kept the reset the same.

That's in blue (Chrome fails) and CSS tree parser complains too:

```
<!DOCTYPE html>
<html>
<head>
	<title>test</title>
</head>
<style type="text/css">

header {
	background-color: blue;
}
@media screen and ((not (monochrome)) or (hover)) {
	header {
		background-color: yellow;
	}
}
</style>
<body>
	<header>
		test header
	</header>
</body>
</html>
```

but separate is in yellow (Chrome is all right) and CSS tree is all right:

```
<!DOCTYPE html>
<html>
<head>
	<title>test</title>
</head>
<style type="text/css">

header {
	background-color: blue;
}
@media not screen and (monochrome), not screen and (hover) {
	header {
		background-color: yellow;
	}
}
</style>
<body>
	<header>
		test header
	</header>
</body>
</html>
```

from what I see it's as if media type has to be always present, plus one can't nest media conditions like example in the spec says...

**Comment by lahmatiy** [comment]
As I realized, there two problems:
- validator web interface was created in time when CSSTree was not tolerant to errors, so any error was fatal. Currently, a parse error is not fatal for entire parsing, bad part are wrapping in Raw nodes instead. However, parse error occur internally and we able to collect them. It should not prevent result output, but it does due to old behavior.
- CSSTree is bad on parsing media queries at the moment. It isn’t support the spec and can parse simple cases only. I’m going to remove Media* nodes and custom parsing in next releases. I believe there must another approach to get a specific structure, and parser should parse in common abstract way as much as possible. Btw, definition syntax can be used to validate an at-rule prelude, like it used for declaration values.

**Comment by revelt** [comment]
@lahmatiy should I close or do you want to keep this open?

**Comment by lahmatiy** [comment]
I think it's better to keep the issue open until it resolved.

**Comment by rgant** [comment]
Is this error in any way related to this ticket?

```
/// Custom breakpoints
/// @access public
/// @param {String} $breakpoint - Breakpoint
/// @param {String} $from - Defaults to "min-width" (mobile first)
///        can also take "max-width" to select elements under the provided size.
@mixin breakpoint($breakpoint, $from: 'min-width') {
  @media ($from: $breakpoint) {
    @content;
  }
}
```

src/styles/abstracts/_mixins.scss
 25:11  ✖  Invalid prelude for `@media`  csstree/validator

**Comment by lahmatiy** [comment]
@rgant Nope, that't stylelint validator issue related to preprocessor's syntax handling. I would be appreciate if you create an issue in stylelint-validator repo. Thanks

**Comment by lahmatiy** [comment]
Fixed in [v3.0.0](https://github.com/csstree/csstree/releases/tag/v3.0.0)
